Compatibility switches

rynowak opened the pull request:
This introduces a pattern for versioning breaking behaviour changes in
minor releases of MVC.

The general plan is that application developers choose a release version
(2.0, 2.1, Latest) as their baseline which determines the effective
'defaults' for some options. Anything the developer sets explicitly is
an override and always wins.

Then we add a version setting to the template to point to the current
release.

This allows us to be progressive with fixing issues and improving areas
that don't work well, but offers the developer some choice about when to
adopt new behaviours. In effect, we separate new behaviours from the
libraries that develiver them. Apps can update the version, and then opt
in to new behaviours as a separate change.

rynowak commented on samples/MvcSandbox/Startup.cs:
```diff
@@ -14,7 +14,7 @@ public class Startup
         public void ConfigureServices(IServiceCollection services)
         {
-            services.AddMvc();
+            services.AddMvc().SetCompatibilityVersion(Microsoft.AspNetCore.Mvc.CompatibilityVersion.Version_2_1);
```
This is just a design PR, I didn't fill in the details or tests yet because I want to get feedback on the overall direction.

pranavkm replied:
Undo

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ConfigureCompatibilityOptions.cs:
```diff
@@ -0,0 +1,86 @@
+    {
+        private readonly ILogger _logger;
+
+        internal ConfigureCompatibilityOptions(
```
This should be public, since we have options classes in other projects in MVC

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs:
```diff
@@ -15,10 +17,15 @@ namespace Microsoft.AspNetCore.Mvc
+        private readonly CompatibilitySwitch<bool> _allowBindingUndefinedValueToEnumType;
+        private readonly CompatibilitySwitch<InputFormatterExceptionModelStatePolicy> _inputFormatterExceptionModelStatePolicy;
+        private readonly CompatibilitySwitch<bool> _suppressJsonDeserializationExceptionMessagesInModelState;
+        private readonly ICompatibilitySwitch[] _switches;
```
Two of the new properties we introduced don't follow the guidelines I wrote up. We need to consider changing these to be more conservative.

Eilon commented:
> when to adopt new behaviours

Such as being British? :trollface:

Eilon commented on src/Microsoft.AspNetCore.Mvc.Core/CompatibilityVersion.cs:
```diff
@@ -0,0 +1,36 @@
+    public enum CompatibilityVersion
+    {
+        /// <summary>
+        /// Sets the default value of settings on <see cref="MvcOptions"/> to match the behaviour of 
```
When this becomes a non-design PR, please switch to American spelling.

Eilon commented on src/Microsoft.AspNetCore.Mvc.Core/CompatibilityVersion.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        /// <summary>
+        /// Sets the default value of settings on <see cref="MvcOptions"/> to match the behaviour of 
+        /// ASP.NET Core MVC 2.0.0.
```
This will sound like a nit, but it isn't. The comments here specify the patch version, but the enum values do not. I think that neither should specify the patch version because that way it will be clearer that this is just about major/minor and not patches. We just won't change such behaviors in patches anyway - except when we do, but when we do, there will be quirks switches for it.

rynowak replied:
Yeah I agree

pranavkm replied:
Would we consider using this mechanism for quirk switches? If we had a way to additionally pick up values for these switches from AppContext, it would make discovering \ using \ testing compat switches consistent both for users and us.

rynowak replied:
There's no reason why we couldn't. I'm not sure if we'll find a case where that adds value though

Eilon replied:
A requirement for quirks switches is that there has to be a way to enable them on .NET Core and .NET Framework *without recompiling* the app. So, if we can come up with something way to do that, it'd be an option.

pranavkm replied:
@Eilon I have to vet the design out, but the idea would be to read from `AppContext.GetData` as an additional source of values for these switches.

Eilon replied:
`AppContext.TryGetSwitch()` is the way to read switches:

https://github.com/Eilon/QuirksModeSwitchSample/blob/master/QuirksModeSwitchSample/Program.cs#L27

But yeah, if we think that's a good feature to have here, that could be reasonable. We'll need to make sure every compat switch has a "reasonable" textual name to use in config files for quirks.

rynowak replied:
I'm not convinced this is a good idea. I don't think these two things serve the same need.

Eilon replied:
@rynowak I agree.

Eilon commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/CompatibilitySwitch.cs:
```diff
@@ -0,0 +1,88 @@
+        private TValue _value; 
+
+        public CompatibilitySwitch(string name)
+            : this(name, default)
```
`defaultValue`?

rynowak replied:
this is a C# 7.1 feature `default` implies `default(TValue)`

Eilon replied:
Oh right!

Eilon commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ConfigureCompatibilityOptions.cs:
```diff
@@ -0,0 +1,86 @@
+            // Evaluate DefaultValues onces to subclasses don't have to cache.
+            var defaultValues = DefaultValues;
+
+            foreach (var @switch in options)
```
Back to actual nits, but if this were named `compatSwitch` everywhere here, it might be easier to read? All this stuff is private/local anyway.

rynowak replied:
Why avoid usage of a language feature for its intended purpose?

Eilon replied:
Meh.

Eilon commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/MvcCompatibilityOptions.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class MvcCompatibilityOptions
+    {
+        public CompatibilityVersion CompatibilityVersion { get; set; } = CompatibilityVersion.Version_2_0;
```
I guess this default value will remain untouched until 3.0, right?

Eilon replied:
If this is true, we should log a bug for the 3.0 milestone to update this value. Otherwise you know we'll forget.

rynowak replied:
👍 Yep. The idea is also that we're drop support for all of the *old* behaviors that we supported in various 2.X release. So the work is more significant than just updating the default, we'd probably obsolete all of the old enum values - I'm not sure yet if its better to remove old enum values or just obsolete them.

Eilon replied:
Ah yes good point! I think we leave them but obsolete them so that we can alert the developer. Otherwise they won't know. Hopefully most people won't even use the switches too long after they migrate.

Eilon commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/MvcOptionsConfigureCompatibilityOptions.cs:
```diff
@@ -0,0 +1,35 @@
+
+                if (Version >= CompatibilityVersion.Version_2_1)
+                {
+                    values[nameof(MvcOptions.InputFormatterExceptionModelStatePolicy)] = InputFormatterExceptionModelStatePolicy.MalformedInputExceptions;
```
Nice, this is already a great use of this feature. cc @kichalla

kichalla commented on src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs:
```diff
@@ -147,11 +147,14 @@ internal static void AddMvcCoreServices(IServiceCollection services)
             services.TryAddEnumerable(
                 ServiceDescriptor.Transient<IConfigureOptions<MvcOptions>, MvcCoreMvcOptionsSetup>());
+            services.TryAddEnumerable(
+                ServiceDescriptor.Transient<IPostConfigureOptions<MvcOptions>, MvcOptionsConfigureCompatibilityOptions>());
```
Just wanted to call this out regarding `Anything the developer sets explicitly is
an override and always wins.`...if a user wants to override one particular setting(like out of 4 switches for example..remember these 4 switches could have also cascaded across multiple minor releases) that a compatibility version is setting, they would need to do something like(which isn't bad)..i.e use [PostConfigure](https://github.com/aspnet/Options/blob/dev/src/Microsoft.Extensions.Options/IPostConfigureOptions.cs#L8) rather than the usual `Configure`
```c#
services.AddMvc().SetCompatibilityVersion(Microsoft.AspNetCore.Mvc.CompatibilityVersion.Version_2_3);
services.PostConfigure<MvcOptions>(o =>
{
    o.AllowBindingUndefinedValueToEnumType = true;
});

```

rynowak replied:
Nah, this is handled here: https://github.com/aspnet/Mvc/pull/7142/files#diff-c9e246ea8e03a841a1b895c97e72dc07R58

If the user manipulates the value it sets `switch.IsValueSet=true` and we don't overwrite it.

kichalla replied:
Ah, I see. Thanks!

rynowak commented:
Updated and removed design tag. I plan to do follow ups to work on the actual settings. This just has tests and infrastructure.

Eilon commented:
@rynowak looks really good. My only current comment is that there are a few missing doc comments on public APIs.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/CompatibilityVersion.cs:
```diff
@@ -0,0 +1,59 @@
+    ///     
+    ///     public void ConfigureServices(IServiceCollection services)
+    ///     {
+    ///         services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_0);
```
I'll change the example to setting this to 2.1. Setting 2.0 is a no-op

rynowak commented:
Rebased and added docs that were missing from new public API surface.

I'm planning on merging this by Friday if I haven't gotten any more important feedback. I'll be doing follow ups to address specific switches as well.

I'm also going to get to work on an announcement for this.

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs:
```diff
@@ -147,11 +147,14 @@ internal static void AddMvcCoreServices(IServiceCollection services)
             services.TryAddEnumerable(
                 ServiceDescriptor.Transient<IConfigureOptions<RouteOptions>, MvcCoreRouteOptionsSetup>());
 
+
```
Unfo

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ConfigureCompatibilityOptions.cs:
```diff
@@ -0,0 +1,105 @@
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            // Evaluate DefaultValues onces to subclasses don't have to cache.
```
once ~~to~~ so

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ICompatibilitySwitch.cs:
```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
```
extra line

pranavkm commented on test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ConfigureCompatibilityOptionsTest.cs:
```diff
@@ -0,0 +1,170 @@
+        {
+            // Arrange
+            var configure = Create(CompatibilityVersion.Version_2_0, new Dictionary<string, object>()
+            {
```
Remove

pranavkm commented on test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ConfigureCompatibilityOptionsTest.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        [Fact]
+        public void Constructor_WithName_IsValueSetIsFalse()
```
Aren't these already covered in `CompatSwitchTest.cs`?

rynowak replied:
Oh yeah, bad copy-paste 👍
